lib/matplotlib/pyplot.py

```diff
@@ -2665,7 +2665,7 @@ def matshow(A: ArrayLike, fignum: None | int = None, **kwargs) -> AxesImage:
     else:
         # Extract actual aspect ratio of array and make appropriately sized
         # figure.
-        fig = figure(fignum, figsize=figaspect(A))
+        fig = figure(fignum, figsize=tuple(*figaspect(A)))
```

Comment:
I suggest instead updating the type hint on `figure` to allow figsize to be an array.

Reply:
The good news is that after fixing the type hint on `plt.figaspect` the CI already failed and warned us about the error. The bad news is that I have no experience and this PR truned it quite the mess, sorry.

So what you suggest would be replacing
```
    figsize: tuple[float, float] | None = None
```
by
```
    figsize: tuple[float, float] | np.ndarray[tuple[Literal[2]], np.dtype[np.floating]] | None = None
```
Or did I get you wrong?

Reply:
I'm afraid I'm not sure about the exact syntax.  Hopefully a typing expert will be along to clarify.

Reply:
I suspect `ArrayLike` in the signature of `Figure` is the best out here as we aggressively normalize the input

```python
        if not np.isfinite(figsize).all() or (np.array(figsize) < 0).any():
            raise ValueError('figure size must be positive finite not '
                             f'{figsize}')
        self.bbox_inches = Bbox.from_bounds(0, 0, *figsize)
```

Reply:
`ArrayLike` is definitely a valid solution, though it feels very broad to me, as it has no shape constrains. Currently, I see 3 options:

1. `ArrayLike`: basically allows almost any input, though many inputs might not be meaningful.
2. `tuple[float, float] | np.ndarray[tuple[Literal[2]], np.dtype[np.floating]]` hinting the reader that we need exact 2 inputs. However, to my knowledge, mypy doesn't check shape anyway.
3. `typle[np.loating, np.floating]`: most strict version, allowing exactly 2 numbers. Changing calls from within to a tuple: `w, h = figsize; plt.figure(figsize=(w, h))`

I think option 2 is the worst one. Options 1 and 2 seem reasonable. I have no strict preference. What do you think @tacaswell?

Reply:
I’d be ok with `ArrayLike`. IMHO typing is a helper to narrow down incorrect use, but one does not have to jump every possible hoop to try and make type annotations exact (which by the way is also not possible in all situations with the current type system).